src/include/encoding.h

```diff
@@ -276,6 +292,13 @@ inline void encode(const buffer::ptr& bp, bufferlist& bl)
   if (len)
     bl.append(bp);
 }
+inline void encode(const buffer::ptr& bp, bufferlist& p_bl, bufferlist& d_bl)
+{
+  __u32 len = bp.length();
+  encode(len, p_bl);
+  if (len)
+    d_bl.append(bp);
```

Comment:
IIUC the main driving factor behind the new primitives is separating _head_ (typically size) from _non-head_ (actual data).
If that's the case, maybe we could reuse the `{encode, decode}_nohead` variants to save us some code duplication.

The obvious problem with applying this on broad scale is about containers and preserving the _head vs no-head_ distinction down the nested encodes. However, do we really need to worry? Isn't the optimization targeting basically several instances of bufferlist known to carry big data?

Reply:
The reason why I ended up creating new primitives is because of the ECSubReadReply structure/message. This needs to encode/decode a map that contains a list that contains pairs that have the data buffers that we want to keep aligned.

<pre>
  std::map<hobject_t, std::list<std::pair<uint64_t, ceph::buffer::list> >> buffers_read;
</pre>

We could write a custom encode/decode method for this structure, duplicating code from encoding.h/denc.h to handle the map, list, pair and buffer::list and call encode/decode_nohead for the data buffers we want to align, but this will end up duplicating code outside of the header files. My viewpoint is that it is better to keep all the encode/decode for containers in one place in the header files, rather than spread custom versions of this through the code, especially as I suspect we will find other places in the code base where this technique can be used to improve data buffer alignment.

Reply:
I see your points. Have you considered an approach with `if constexpr()` and a _head-nohead trait_ (a template parameter; we already make use of the _traits_ concept in the dencoding frameworks) passed down the stack? It would require two passes (one with matadata bl, one with data bl) but maybe we could simplify even already existing code (by streamlining `{de,en}code_nohead()`).

Just thinking loudly.

Reply:
I really liked that idea and did a bit of prototyping. My observations:

1. The extra if statements in the {de,en}code functions are a bit messy, but perhaps 1 template with some ifs is still better than 2 templates for each type
2. I could write generic templates for encode_nohead, encode_headonly, encode that pass a different trait to the encode templates which means some of the nohead templates could be deleted
3. Decoding doesn't work :-( For compound types like a map you can't do the decode in two passes, you need to interleave the decode of the two buffers. This is also why decode_nohead requires extra parameters.

But the prototyping taught me that I don't need to duplicate all the templates - I can just have {de,en}code templates for types that take 2 bufferlists for payload and data buffers and then write a generic template like this to continue supporting encoding with a single bufferlist:

`template<class T> inline void encode(T& t,bufferlist &bl) { encode(t,bl,bl); }
`

This will tidy things up a lot and I think achieves what you were looking for. Will do this for the next commit